### PR TITLE
[WIP] Building blocks for PEBBLE

### PR DESCRIPTION
**Pull request by dan-pandori**
## Description

Creates an entropy reward replay wrapper to support the unsupervised state entropy based pre-training of an agent, as described in the PEBBLE paper.
https://sites.google.com/view/icml21pebble

## Testing

Added unit tests.

### CODE REVIEW

File: src/imitation/util/networks.py [hunk 1]
```diff
@@ -126,12 +129,12 @@ def update_stats(self, batch: th.Tensor) -> None:
         tot_count = self.count + batch_count
         self.running_mean += delta * batch_count / tot_count
 
-        self.running_var *= self.count
-        self.running_var += batch_var * batch_count
-        self.running_var += th.square(delta) * self.count * batch_count / tot_count
-        self.running_var /= tot_count
+        m_a = self.running_var * self.count
```
Comment (AdamGleave) [inline src/imitation/util/networks.py:132]: @taufeeque9 does this rewrite look OK for you from a numerical stability perspective?

File: src/imitation/util/util.py [hunk 2]
```diff
@@ -359,3 +359,34 @@ def get_first_iter_element(iterable: Iterable[T]) -> Tuple[T, Iterable[T]]:
...
+        all_obs: The tensor of all states to compare to.
+        k: the number of neighbors to consider
+
+    Returns:
+        A tensor containing the state entropy for `obs`.
+    """
+    assert obs.shape[1:] == all_obs.shape[1:]
+    with th.no_grad():
+        non_batch_dimensions = tuple(range(2, len(obs.shape) + 1))
+        distances_tensor = th.linalg.vector_norm(
```
Comment (AdamGleave) [inline src/imitation/util/util.py:382]: This tensor could get quite big O(batch_size * total_timesteps * unsupervised_pretraining_frac). Don't want to premature optimize, but we should do some testing before merging to see if this is a problem in practice, I'm sure there's some more efficient ways to find KNN distances.

File: src/imitation/algorithms/preference_comparisons.py [hunk 3]
```diff
@@ -316,6 +330,43 @@ def logger(self, value: imit_logger.HierarchicalLogger) -> None:
...
+            raise ValueError(
+                f"{self.__class__.__name__} expects "
+                f"{PebbleStateEntropyReward.__name__} reward function",
+            )
+        super().__init__(reward_fn=reward_fn, **kwargs)
+
+    def unsupervised_pretrain(self, steps: int, **kwargs: Any) -> None:
+        self.train(steps, **kwargs)
+        fn = self.reward_fn
+        fn.unsupervised_exploration_finish()  # type: ignore[attribute-error]
```
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:367]: Odd that the `type: ignore` is needed given the earlier type annotation, does adding an explicit `assert isinstance(fn, PebbleStateEntropyReward)` clear it?

File: src/imitation/algorithms/preference_comparisons.py [hunk 4]
```diff
@@ -1670,25 +1726,30 @@ def train(
...
-        shares = util.oric(probs * total_comparisons)
-        schedule = [initial_comparisons] + shares.tolist()
-        print(f"Query schedule: {schedule}")
-
-        timesteps_per_iteration, extra_timesteps = divmod(
-            total_timesteps,
-            self.num_iterations,
-        )
+        preference_query_schedule = self._preference_gather_schedule(total_comparisons)
+        print(f"Query schedule: {preference_query_schedule}")
```
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:1731]:
```suggestion
        self.logger.log(f"Query schedule: {preference_query_schedule}")
```

File: src/imitation/algorithms/preference_comparisons.py [hunk 5]
```diff
@@ -1670,25 +1726,30 @@ def train(
...
         reward_loss = None
         reward_accuracy = None
 
-        for i, num_pairs in enumerate(schedule):
+        ###################################################
+        # Pre-training agent before gathering preferences #
+        ###################################################
+        with self.logger.accumulate_means("agent"):
+            self.logger.log(
+                f"Pre-training agent for {unsupervised_pretrain_timesteps} timesteps",
```
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:1746]: Consider skipping this and next line if `unsupervised_pretrain_timesteps` is zero? Seems a bit superflous.

**Review by AdamGleave — COMMENTED** [review]
Initial review, I haven't looked at the tests yet.

File: src/imitation/util/util.py (hunk 2 above)
Comment (mifeet) [reply src/imitation/util/util.py:382]: Yeah, the original implementation actually does this by parts: https://github.com/rll-research/BPref/blob/f3ece2ecf04b5d11b276d9bbb19b8004c29429d1/agent/sac.py#L13

Shall we implement something similar?

File: src/imitation/util/networks.py (hunk 1 above)
Comment (mifeet) [reply src/imitation/util/networks.py:132]: The computation should be identical to the original, it's just that running_var is not updated in place. This shouldn't change the result unless I miss something?

Anyway, this change isn't necessary, it can be reverted.

File: src/imitation/scripts/config/train_preference_comparisons.py [hunk 6]
```diff
@@ -60,6 +61,31 @@ def train_defaults():
...
+    rl = {
+        "rl_cls": sb3.SAC,
+        "batch_size": 256,  # batch size for RL algorithm
+        "rl_kwargs": {"batch_size": None},  # make sure to set batch size to None
+    }
+    train = {
+        "policy_cls": base.SAC1024Policy,  # noqa: F841
+    }
+    common = {"env_name": "MountainCarContinuous-v0"}
+    allow_variable_horizon = True
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/scripts/config/train_preference_comparisons.py:85]: I understand, but this is related to the MountainCarContinuous-v0 environment. As far as I can see, the same pattern is used in other places (e.g., line 93). 
Is there a way to avoid this without using a different environment? I noticed there are things like `AbsorbAfterDoneWrapper` but I didn't dig deeper.

File: src/imitation/policies/replay_buffer_wrapper.py [hunk 7]
```diff
@@ -61,6 +85,8 @@ def __init__(
         self.reward_fn = reward_fn
         _base_kwargs = {k: v for k, v in kwargs.items() if k in ["device", "n_envs"]}
         super().__init__(buffer_size, observation_space, action_space, **_base_kwargs)
+        if isinstance(reward_fn, ReplayBufferAwareRewardFn):
+            reward_fn.on_replay_buffer_initialized(self)
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/policies/replay_buffer_wrapper.py:89]: It wasn't the intention, but may be nicer. We would need to propagate observations shape as well.

File: src/imitation/algorithms/pebble/__init__.py [hunk 8]
```diff
@@ -0,0 +1 @@
+"""PEBBLE specific algorithms."""
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/pebble/__init__.py:1]: We can do that, e.g., classes for work with fragments and preference gathering seem like independent pieces of logic. Probably for another PR, though.

File: src/imitation/algorithms/pebble/entropy_reward.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
+"""Reward function for the PEBBLE training algorithm."""
+
+from enum import Enum, auto
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/pebble/entropy_reward.py:3]: Fixed

File: src/imitation/algorithms/pebble/entropy_reward.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+
+class PebbleRewardPhase(Enum):
+    """States representing different behaviors for PebbleStateEntropyReward."""
+
+    UNSUPERVISED_EXPLORATION = auto()  # Entropy based reward
+    POLICY_AND_REWARD_LEARNING = auto()  # Learned reward
+
+
+class PebbleStateEntropyReward(ReplayBufferAwareRewardFn):
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/pebble/entropy_reward.py:26]: No intention here, this was a natural result of the refactoring out of the original buffer wrapper.

I agree separating the two would be nice, and it would be easier to test. Two things complicate it, though:
* when there are not enough observations, we would need to fall back to something, eg., zero vector (which is probably fine)
* it complicates pickling. Either the entropy part would be aware of state changes with ReplayBufferView and the setting logic would be duplicated in both, or unaware but then we may loose entropy_stats during pickling (it would be easier to explain in code)

I may take a look later if there is time.

File: src/imitation/algorithms/pebble/entropy_reward.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+
+        Args:
+            learned_reward_fn: The learned reward function used after unsupervised
+                exploration is finished
+            nearest_neighbor_k: Parameter for entropy computation (see
+                compute_state_entropy())
+        """
+        self.learned_reward_fn = learned_reward_fn
+        self.nearest_neighbor_k = nearest_neighbor_k
+        self.entropy_stats = RunningNorm(1)
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/pebble/entropy_reward.py:61]: We could, and I can do it if you insist, but I would prefer not to, to keep things simple.

NormalizedRewardNet is on the 4th level of class inheritance and there's a lot going on. Unless someone knows precisely what it does it may be confusing. I think the little functionality duplication is worth the simplicity.

File: src/imitation/algorithms/pebble/entropy_reward.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+        self.state = PebbleRewardPhase.UNSUPERVISED_EXPLORATION
+
+        # These two need to be set with set_replay_buffer():
+        self.replay_buffer_view: Optional[ReplayBufferView] = None
+        self.obs_shape: Union[Tuple[int, ...], Dict[str, Tuple[int, ...]], None] = None
+
+    def on_replay_buffer_initialized(self, replay_buffer: ReplayBufferRewardWrapper):
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/pebble/entropy_reward.py:68]: When designing the API of `ReplayBufferAwareRewardFn`, it seemed more natural and decoupled to have a signature of 

    on_replay_buffer_initialized(self, replay_buffer: ReplayBufferRewardWrapper)

rather than

    on_replay_buffer_initialized(self, replay_buffer_view: ReplayBufferView, obs_shape: Union[Tuple[int, ...], Dict[str, Tuple[int, ...]]])

since potential future replay buffer aware functions may need something other than `obs_shape` as well. It's a minor reason, though.

File: src/imitation/algorithms/preference_comparisons.py [hunk 13]
```diff
@@ -75,6 +77,18 @@ def sample(self, steps: int) -> Sequence[TrajectoryWithRew]:
             be the environment rewards, not ones from a reward model).
         """  # noqa: DAR202
 
+    def unsupervised_pretrain(self, steps: int, **kwargs: Any) -> None:
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/preference_comparisons.py:92]: Oh, right, I didn't realize that. 
It complicates things a bit, but I've pushed a possible solution with a check and warning.

File: src/imitation/algorithms/preference_comparisons.py (hunk 3 above)
Comment (mifeet) [reply src/imitation/algorithms/preference_comparisons.py:385]: Fixed

File: src/imitation/algorithms/preference_comparisons.py (hunk 5 above)
Comment (mifeet) [reply src/imitation/algorithms/preference_comparisons.py:1764]: Makes sense, fixed

File: src/imitation/scripts/config/train_preference_comparisons.py [hunk 14]
```diff
@@ -60,6 +61,31 @@ def train_defaults():
...
+
+    rl = {
+        "rl_cls": sb3.SAC,
+        "batch_size": 256,  # batch size for RL algorithm
+        "rl_kwargs": {"batch_size": None},  # make sure to set batch size to None
+    }
+    train = {
+        "policy_cls": base.SAC1024Policy,  # noqa: F841
+    }
+    common = {"env_name": "MountainCarContinuous-v0"}
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/scripts/config/train_preference_comparisons.py:84]: One could argue that could also happen with `with pebble some_rl_class` vs `with some_rl_class pebble`, but I get you're point, that's much more likely to happen.

I've extracted the environment from `pebble` configuration. If no env is specified, the error message  is pretty self-explanatory (`The algorithm only supports ... as action spaces but ... was provided`). 

Not sure, how to improve it with config hooks. One could change the environment if pebble is activated and default env is set, but it seems a bit magical.

File: src/imitation/scripts/config/train_preference_comparisons.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/scripts/config/train_preference_comparisons.py]: Resolved together with the point above.

File: src/imitation/util/util.py (hunk 2 above)
Comment (mifeet) [reply src/imitation/util/util.py]: Ok, I went ahead and pushed the batched version.

File: src/imitation/rewards/reward_function.py [hunk 15]
```diff
@@ -32,3 +32,5 @@ def __call__(
         Returns:
             Computed rewards of shape `(batch_size,`).
         """  # noqa: DAR202
+
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/rewards/reward_function.py]: Resolved

File: src/imitation/algorithms/pebble/entropy_reward.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/pebble/entropy_reward.py:61]: I'd prefer wrapping it with a `NormalizedRewardNet`, they're conceptually doing very different things, and we might want to use different normalization schemes (`RunningNorm` often works worse than `EMANorm`)

File: src/imitation/algorithms/pebble/entropy_reward.py [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+        all_observations = self.replay_buffer_view.observations
+        # ReplayBuffer sampling flattens the venv dimension, let's adapt to that
+        all_observations = all_observations.reshape((-1, *self.obs_shape))
+
+        if all_observations.shape[0] < self.nearest_neighbor_k:
+            # not enough observations to compare to, fall back to the learned function;
+            # (falling back to a constant may also be ok)
+            return self.learned_reward_fn(state, action, next_state, done)
+        else:
+            # TODO #625: deal with the conversion back and forth between np and torch
```
Comment (AdamGleave) [inline src/imitation/algorithms/pebble/entropy_reward.py:109]: is this still an issue?

File: src/imitation/algorithms/pebble/entropy_reward.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/pebble/entropy_reward.py:26]: Let's leave it as is for now given it'd be a fairly major change, but please don't resolve this comment (I might ask someone else to look at making the change before merging, or merge but make an issue to track it).

File: src/imitation/algorithms/preference_comparisons.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:95]: thanks, this seems like a good solution!

File: src/imitation/algorithms/pebble/entropy_reward.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/pebble/entropy_reward.py:69]: It made sense in tests, but I used a different solution and removed `set_replay_buffer()` as suggested.

File: src/imitation/algorithms/preference_comparisons.py [hunk 17]
```diff
@@ -75,6 +77,40 @@ def sample(self, steps: int) -> Sequence[TrajectoryWithRew]:
...
+        Returns:
+            True if this generator has a pre-training phase, False otherwise
+        """
+        orig_impl = TrajectoryGenerator.unsupervised_pretrain
+        return type(self).unsupervised_pretrain != orig_impl
+
+    def unsupervised_pretrain(self, steps: int, **kwargs: Any) -> None:
+        """Pre-train an agent before collecting comparisons.
+
+        By default, this method asserts that pre-training has zero steps allocated.
```
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/preference_comparisons.py]: Fixed

File: src/imitation/policies/replay_buffer_wrapper.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/policies/replay_buffer_wrapper.py:89]: See https://github.com/HumanCompatibleAI/imitation/pull/625/files#r1037735386

**Comment by mifeet** [comment]
@AdamGleave: reacting to your comments here together:

> I'd prefer wrapping it with a NormalizedRewardNet, they're conceptually doing very different things, and we might want to use different normalization schemes (RunningNorm often works worse than EMANorm)

Ok, it required a larger refactor, but you can see how it looks in the [last couple of commits](https://github.com/HumanCompatibleAI/imitation/compare/c80fb80109647e193d7afde2c0ef3bdcabf9dcfc...74ba96b17d75c785ad5a8af6725f645dbd17df9e). 

A good thing is that this change also addresses your [other comment](https://github.com/HumanCompatibleAI/imitation/pull/625/files#r1037724606). It simplified the entropy reward classes (separate entropy reward and switching from pre-traininig reward) and allows for more configurability, at the expense at making wiring a little more complicated.

It also results in two changes internally:
* Previously, the running mean/var statistics for normalization were first updated, then normalization was applied. Now these are swapped.
* Previously, reward calculation required conversions numpy -> torch -> numpy, now it internally converts numpy -> torch -> numpy -> torch -> numpy (because that's what the existing code for NormalizedRewardNet does).

File: src/imitation/algorithms/pebble/entropy_reward.py (hunk 16 above)
Comment (mifeet) [reply src/imitation/algorithms/pebble/entropy_reward.py]: See https://github.com/HumanCompatibleAI/imitation/pull/625#issuecomment-1345385184

File: src/imitation/algorithms/pebble/entropy_reward.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mifeet) [reply src/imitation/algorithms/pebble/entropy_reward.py]: See https://github.com/HumanCompatibleAI/imitation/pull/625#issuecomment-1345385184
